Purchases/Networking/Operations/PostReceiptDataOperation.swift

```diff
@@ -0,0 +1,85 @@
+
+class PostReceiptDataOperation: NetworkOperation {
+
+    private let customerInfoCallbackCache: CallbackCache<CustomerInfoCallback>
+    let httpClient: HTTPClient
+    let authHeaders: [String: String]
+
+    init(httpClient: HTTPClient,
+         authHeaders: [String: String],
+         customerInfoCallbackCache: CallbackCache<CustomerInfoCallback>) {
+        self.httpClient = httpClient
+        self.authHeaders = authHeaders
+        self.customerInfoCallbackCache = customerInfoCallbackCache
+    }
+
+    // swiftlint:disable:next function_parameter_count
+    func post(receiptData: Data,
+              appUserID: String,
+              isRestore: Bool,
+              productInfo: ProductInfo?,
+              presentedOfferingIdentifier offeringIdentifier: String?,
+              observerMode: Bool,
+              subscriberAttributes subscriberAttributesByKey: SubscriberAttributeDict?,
+              completion: @escaping BackendCustomerInfoResponseHandler) {
+        let fetchToken = receiptData.asFetchToken
+        var body: [String: Any] = [
+            "fetch_token": fetchToken,
+            "app_user_id": appUserID,
+            "is_restore": isRestore,
+            "observer_mode": observerMode
+        ]
+
+        let cacheKey =
+        """
+        \(appUserID)-\(isRestore)-\(fetchToken)-\(productInfo?.cacheKey ?? "")
+        -\(offeringIdentifier ?? "")-\(observerMode)-\(subscriberAttributesByKey?.debugDescription ?? "")"
+        """
+
+        let callbackObject = CustomerInfoCallback(key: cacheKey, callback: completion)
+        if customerInfoCallbackCache.add(callback: callbackObject) == .addedToExistingInFlightList {
+            return
+        }
+
+        if let productInfo = productInfo {
+            body.merge(productInfo.asDictionary()) { _, new in new }
+        }
+
+        if let subscriberAttributesByKey = subscriberAttributesByKey {
+            let attributesInBackendFormat = subscriberAttributesToDict(subscriberAttributes: subscriberAttributesByKey)
+            body["attributes"] = attributesInBackendFormat
+        }
+
+        if let offeringIdentifier = offeringIdentifier {
+            body["presented_offering_identifier"] = offeringIdentifier
+        }
+
+        httpClient.performPOSTRequest(serially: true,
+                                      path: "/receipts",
+                                      requestBody: body,
+                                      headers: authHeaders) { statusCode, response, error in
+            self.customerInfoCallbackCache.performOnAllItemsAndRemoveFromCache(withKey: cacheKey) { callbackObject in
+                self.handle(customerInfoResponse: response,
+                            statusCode: statusCode,
+                            maybeError: error,
+                            completion: callbackObject.callback)
+            }
+        }
+    }
+}
+
+extension PostReceiptDataOperation: SubscriberAttributesMarshalling, CustomerInfoResponseHandling { }
```

Comment:
Again, here we should **_probably_** inject these things instead of having them be protocols.

Reply:
Moved to injection.